### PR TITLE
refactor: waitNext map as a type

### PR DESCRIPTION
**Pull request by esadakar**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1863)

### CODE REVIEW

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -856,13 +831,8 @@ func (k *Kad) Disconnected(peer p2p.Peer) {
...
 	k.connectedPeers.Remove(peer.Address, po)
 
-	k.waitNextMu.Lock()
-	newInfo := retryInfo{tryAfter: time.Now().Add(timeToRetry), failedAttempts: 0}
-	if info, ok := k.waitNext[peer.Address.ByteString()]; ok {
-		newInfo.failedAttempts = info.failedAttempts
-	}
-	k.waitNext[peer.Address.ByteString()] = newInfo
-	k.waitNextMu.Unlock()
+	failedAttemts := k.waitNext.Attempts(peer.Address)
```
Comment (ldeffenb) [inline pkg/topology/kademlia/kademlia.go:834]: Attemts -> Attempts

File: pkg/topology/kademlia/kademlia.go [hunk 2]
```diff
@@ -859,13 +828,8 @@ func (k *Kad) Disconnected(peer p2p.Peer) {
...
 
-	k.waitNextMu.Lock()
-	newInfo := retryInfo{tryAfter: time.Now().Add(timeToRetry), failedAttempts: 0}
-	if info, ok := k.waitNext[peer.Address.ByteString()]; ok {
-		newInfo.failedAttempts = info.failedAttempts
-	}
-	k.waitNext[peer.Address.ByteString()] = newInfo
-	k.waitNextMu.Unlock()
+	failedAttempts := k.waitNext.Attempts(peer.Address)
+	k.waitNext.Set(peer.Address, time.Now().Add(timeToRetry), failedAttempts)
```
Comment (janos) [inline pkg/topology/kademlia/kademlia.go:832]: This is actually, not an atomic operation. I would suggest to create a new method on WaitNext just to set the time and leave the counter as it is under a single lock.

**Review by janos — COMMENTED** [review]
Very nice. Thanks for doing this @esadakar.

File: pkg/topology/kademlia/internal/waitnext/waitnext.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
+package waitnext
```
[the remark this answers is not in the document]
Comment (esadakar) [reply pkg/topology/kademlia/internal/waitnext/waitnext.go:7]: done

File: pkg/topology/kademlia/internal/waitnext/waitnext_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package waitnext_test
```
[the remark this answers is not in the document]
Comment (esadakar) [reply pkg/topology/kademlia/internal/waitnext/waitnext_test.go:7]: done

**Review by anatollupacescu — APPROVED** [review]

**Review by janos — APPROVED** [review]
Great, thanks!

**Pull request merged**